file-store/src/main/java/org/keycloak/models/map/storage/file/writer/YamlWritingMechanism.java

```diff
@@ -0,0 +1,149 @@
+import java.util.Optional;
+import java.util.function.Consumer;
+import org.snakeyaml.engine.v2.common.FlowStyle;
+import org.snakeyaml.engine.v2.common.ScalarStyle;
+import org.snakeyaml.engine.v2.events.DocumentEndEvent;
+import org.snakeyaml.engine.v2.events.DocumentStartEvent;
+import org.snakeyaml.engine.v2.events.Event;
+import org.snakeyaml.engine.v2.events.ImplicitTuple;
+import org.snakeyaml.engine.v2.events.MappingEndEvent;
+import org.snakeyaml.engine.v2.events.MappingStartEvent;
+import org.snakeyaml.engine.v2.events.ScalarEvent;
+import org.snakeyaml.engine.v2.events.SequenceEndEvent;
+import org.snakeyaml.engine.v2.events.SequenceStartEvent;
+import org.snakeyaml.engine.v2.events.StreamEndEvent;
+import org.snakeyaml.engine.v2.events.StreamStartEvent;
+import org.snakeyaml.engine.v2.nodes.Tag;
+
+/**
+ * Mechanism which produces list of {@link Event}s. 
+ * 
+ * @author vramik
+ */
+public class YamlWritingMechanism implements WritingMechanism, Closeable {
+
+    private final ImplicitTuple implicitTuple = new ImplicitTuple(true, true);
+    private final Consumer<Event> consumer;
+    private boolean runningPreTasks = false;
+    private final LinkedList<RunOnlyOnce> preTasks = new LinkedList<RunOnlyOnce>() {
+        @Override
+        public RunOnlyOnce removeLast() {
+            final RunOnlyOnce res = super.removeLast();
+            res.runPostTask();
+            return res;
+        }
+    };
+
+    public YamlWritingMechanism(Consumer<Event> consumer) {
+        this.consumer = consumer;
+        this.preTasks.add(new RunOnlyOnce(this::startDocument, this::endDocument));
+    }
+
+    @Override
+    public void close() {
+        endDocument();
+    }
+
+    public YamlWritingMechanism startMapping() {
+        this.consumer.accept(new MappingStartEvent(Optional.empty(), Optional.of(Tag.MAP.getValue()), true, FlowStyle.BLOCK));
+        return this;
+    }
+
+    public YamlWritingMechanism endMapping() {
+        this.consumer.accept(new MappingEndEvent());
+        return this;
+    }
+
+    private YamlWritingMechanism writeObject(Runnable taskWithOptionalWrite, Runnable preWriteTask, Runnable postWriteTask) {
+        RunOnlyOnce roo = new RunOnlyOnce(preWriteTask, postWriteTask);
+        try {
+            preTasks.addLast(roo);
+            taskWithOptionalWrite.run();
+        } finally {
+            preTasks.removeLast();
+        }
+        return this;
+    }
+
+
+    @Override
+    public YamlWritingMechanism writeMapping(Runnable task) {
+        return writeObject(task, this::startMapping, this::endMapping);
+    }
+
+    @Override
+    public YamlWritingMechanism writeSequence(Runnable task) {
+        return writeObject(task, this::startSequence, this::endSequence);
+    }
+
+    @Override
+    public YamlWritingMechanism writePair(String key, Runnable task) {
+        return writeObject(task, () -> writeObject(key), null);
+    }
+
+    public YamlWritingMechanism startSequence() {
```

Comment:
I believe it makes sense to make it private as `WritingMechanism` doesn't contain it any more and the method is used only from within this class. wdyt?
```suggestion
    private YamlWritingMechanism startSequence() {
```